ipc_sendfile.c

```diff
@@ -0,0 +1,185 @@
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ipc.h>
+#include <sys/msg.h>
+#include <mqueue.h>
+#include <fcntl.h>
+#include <limits.h>
+
+//#include "ipc_method.h"
+
+#define MQ_MAXMSG 10
+#define MQ_MSGSIZE 8192
+
+long int find_file_size(long int fd);
+
+int main(int argc, char *argv[])
+{
+	int opt;
+	int fflag = 0;
+	int qflag = 0;
+	int pflag = 0;
+	int sflag = 0;
+	char method_name[50]={};
+	char file_name[50]={};
+	// file size
+	long int fd;
+	long int file_size=0;
+	int error_check=0;
+
+	if (argc < 2)
+		printf("Missing arguments or extra arguments, use -h for more information.\n");
+
+
+	while ((opt = getopt(argc, argv, "hf:q:p:s:")) != -1) {
+		switch (opt) {
+			case 'h':
+				printf("<HELP>\n"
+					"-q <queue_name>, use queue as IPC method\n"
+					"-p <pipe_name>, uses pipes as IPC method\n"
+					"-s <buffer size in kB>, uses shared memory as IPC method\n"
+					"-f <filename>, file used to read/write data\n"
+					"</HELP>\n");
+			break;
+			case 'q':
+				qflag ++;
+				strcpy(method_name, optarg);
+				
+
+				break;
+			case 'p':
+				pflag ++;
+				strcpy(method_name, optarg);
+				break;
+			case 's':
+				sflag ++;
+				strcpy(method_name, optarg);
+				break;
+			case 'f':
+				fflag ++;
+				strcpy(file_name, optarg);
+
+				fd = open(file_name, O_RDONLY,0660);
+				file_size = find_file_size(fd);
+				if (fd == -1) {
+					printf("file name: %s, can not open \n", file_name);
+					perror("open");
+					exit(EXIT_FAILURE);
+				}
+				break;
+			default: /* '?' */
+				fprintf(stderr, "Usage: \n"
+						"-h <--help print out containing description of all supported command-line arguments>\n");
+				exit(EXIT_FAILURE);
+
+		}
+	}
+        
+	if (fflag == 1) {
+		if ( pflag + qflag + sflag != 1) {
+			printf("name of file is used: %s\n", file_name);
+			printf("Select one IPC method!e.g \n"
+					"./ipcSend -q /<queue_name> -f <file_name>\n"
+					"./ipcSend -p <pipe_name> -f <file_name>\n"
+					"./ipcSend -s <shared_name> -f <file_name>\n");
+			exit(EXIT_FAILURE);
+
+		} else {
+			if(pflag){
+				printf("<pipe is used, pipe name is %s>\n", method_name);
+				printf("Send file name: %s\n", file_name);
+			} else if (qflag) {
+				printf("<queue is used, queue name is %s>\n", method_name);
+				printf("Send file name: %s\n", file_name);
+				
+				//message queue descriptors
+				mqd_t mq;
+				
+				// open a message queue using default attributes
+				mq = mq_open(method_name, O_CREAT | O_EXCL | O_WRONLY, 0666, NULL);
+				
+				if (mq == -1)
+				{
+					perror("mq_open");
+					exit(EXIT_FAILURE);
+				}
+					
+				long int send = 0;
+				char *buffer = malloc(MQ_MSGSIZE);
+				
+				while (send < file_size)
```

Comment:
An alternative way to detect the end of the file, which might improve readability by reducing the number of variables is that if you read all the file contents the next call to `read` will return 0, so this would work too:

```suggestion
				while (read(fd, buffer, MQ_MSGSIZE) != 0)
```

Reply:
If instead you want to save the return code in order to account for error handling you could also do:
```suggestion
				while ((ret_code = read(fd, buffer, MQ_MSGSIZE)) != 0)
```

Reply:
Your idea í right, however, if I use ret_code = read(fd, buffer, MQ_MSGSIZE, I still need to declare a variable to get total_send_size so I keep the way I did.

Reply:
An assignment operation returns the value of the left operand after the operation https://stackoverflow.com/questions/9514569/what-does-an-assignment-return